strictdoc/export/html/html_templates.py

```diff
@@ -17,13 +17,31 @@
     nodes,
 )
 from jinja2.ext import Extension
+from markupsafe import Markup
 
 from strictdoc import environment
 from strictdoc.core.project_config import ProjectConfig
 from strictdoc.helpers.file_modification_time import get_file_modification_time
 from strictdoc.helpers.timing import measure_performance
 
 
+class JinjaEnvironment:
+    environment: Environment
+
+    def __init__(self, environment: Environment):
+        self.environment = environment
+
+    def get_template(self, *args, **kwargs):
+        return self.environment.get_template(*args, **kwargs)
+
+    def render_template_as_markup(
```

Comment:
I think we are very fine with adding this method to control what is passed to or taken from the render() call.

One thing that I am slightly missing to understand:

Does it mean that this solution replaces the need to do `autoescape=True` because you anyway wrap the output into Markup?

Or the `autoescape=True` is still needed?

Reply:
Short comment only as I'm afk. Yes autoescape ist needed. It works like this:

1. autoescape escapes every object that comes into the template from the Python side
2. except if the object is of type markupsafe.Markup, which tells autoescape "Don't touch me"

The Markup type is a very str compatible wrapper around str. isinstance(Markup("Test"), str) is True. Wrapping a str into Markup does NOT escape the str. For more details see the markupsafe project.

Does this address your question?

Reply:
Your answer addresses my question to 80% but the remaining 20% is something that I see contradicting, given the experiment below.

To understand the behavior of Jinja and markupsafe I went to experiment with you branch a little bit and found something that I don't understand given your answer.

See this commit and the comment that I made:

```
{# WIP: DO NOT UNDERSTAND WHY adding |safe here makes the section title to be escaped #}
```

https://github.com/strictdoc-project/strictdoc/commit/66a380bdca13bd699969d1c8c6297b41802e5036

I cannot understand why the new itest passes when the `|safe` is there, i.e., the section does get escaped, and the other way round - if I remove `|safe`, it is NOT escaped.

My understanding is that `safe` means exactly the opposite of `do escape` because it says `do not escape me, I am escaped already`.

What am I missing?

Reply:
I think the key to understand it is [another property](https://markupsafe.palletsprojects.com/en/latest/escaping/#markupsafe.Markup) of the `Markup` type.

> This [the Markup type] is a subclass of [str](https://docs.python.org/3/library/stdtypes.html#str). It has the same methods, *but escapes their arguments*

Consider this minimal code:

```python
from markupsafe import Markup

Markup("Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ") + "<div>"
Markup("To prevent it, mark args as safe: ") + Markup("<div>")
```

In your example, you set `autoescape=False`. So expectation is that title would not be escaped. But when we apply the `| safe` filter to `".&nbsp;"`, we create a `Markup(".&nbsp;")` object. If we add the title (a raw `str`) to it, `Markup.__add__` will escape `title` during the `+` operation.

Does this make sense?

Reply:
Now it makes complete sense. I just tried it this way, and this is even closer to the line in my commit:

```
>>> "<div>" + Markup("Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ")
Markup("&lt;div&gt;Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ")
```

I kind of felt yesterday that it should work this way but didn't go to check it myself, having asked my question above instead. Thanks for explaining!

----

Could you also explain how can it work in both directions? I just tried this naive thing and it is not working the way Markup does above.

```py
class GFG(str):

    def __init__(self, val):
        self.val = val

    def __add__(self, val2):
        return GFG(self.val + val2.val)


obj1 = "GEEKS"
obj2 = GFG("ForGeeks")
obj3 = obj1 + obj2
print(obj3.val)
```

```
Traceback (most recent call last):
  File "/home/stanislav/sandbox/volatile/pythonProject2/main.py", line 13, in <module>
    print(obj3.val)
AttributeError: 'str' object has no attribute 'val'
```

----


I understand the behavior now perfectly, and I am now wondering if there is a way to improve the readability of all affected lines such that one does not have to mentally do this rule calculation. I.e.,

```
{%- set field_content_ = field_content_ + sdoc_entity.context.title_number_string + ".&nbsp;"|safe -%}
```

Written this way, one really has to be trained into understanding that:

1) `|safe` only works on the immediate member, not the whole string.
2) there is `[another property](https://markupsafe.palletsprojects.com/en/latest/escaping/#markupsafe.Markup) of the Markup type` like you explained.

We are fine to go with all this written as-is in your PR. I just feel like these things are not explicit enough and somewhat increase the level of magic in the Jinja templates.

Reply:
> Could you also explain how can it work in both directions?

Could you explain what you mean with "both directions"?

> I just tried this naive thing and it is not working the way Markup does above.

The `Markup` class doesn't aggregate a `str` attribute to hold the value. It rather *is* the value. This is accomplished by calling `str.__new__` from `Markup.__new__`. See [their implementation](https://github.com/pallets/markupsafe/blob/main/src/markupsafe/__init__.py#L81).

> |safe only works on the immediate member, not the whole string

What is the immediate member?

> I just feel like these things are not explicit enough and somewhat increase the level of magic in the Jinja templates.

I see your point. Yes it's a bit magic and I have no good idea how to make it more explicit without loosing the advantages of being less explicit. Besides adding comments and referring to documentation ([Autoescaping](https://jinja.palletsprojects.com/en/3.1.x/api/#autoescaping), [Working with Automatic Escaping](https://jinja.palletsprojects.com/en/3.1.x/templates/#working-with-automatic-escaping), [Working With Safe Text](https://markupsafe.palletsprojects.com/en/2.1.x/escaping/#markupsafe.Markup)). That could be considered lame though :)

Reply:
> > Could you also explain how can it work in both directions?
> 
> Could you explain what you mean with "both directions"?

The order of Markup and str doesn't matter. This works:

```
>>> Markup("Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ") + "<div>"
Markup("Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: &lt;div&gt;")
```

and this works:

```
>>> "<div>" + Markup("Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ")
Markup("&lt;div&gt;Markup's str functions (e.g. __add__) consider raw str args as unsafe and escape them: ")
```

My naive example (see my previous comment) doesn't allow the other direction, i.e., when the `str` goes first and it is added with Markup.

> > I just tried this naive thing and it is not working the way Markup does above.
> 
> The `Markup` class doesn't aggregate a `str` attribute to hold the value. It rather _is_ the value. This is accomplished by calling `str.__new__` from `Markup.__new__`. See [their implementation](https://github.com/pallets/markupsafe/blob/main/src/markupsafe/__init__.py#L81).
> 
> > |safe only works on the immediate member, not the whole string
> 
> What is the immediate member?
> 
> > I just feel like these things are not explicit enough and somewhat increase the level of magic in the Jinja templates.
> 
> I see your point. Yes it's a bit magic and I have no good idea how to make it more explicit without loosing the advantages of being less explicit. Besides adding comments and referring to documentation ([Autoescaping](https://jinja.palletsprojects.com/en/3.1.x/api/#autoescaping), [Working with Automatic Escaping](https://jinja.palletsprojects.com/en/3.1.x/templates/#working-with-automatic-escaping), [Working With Safe Text](https://markupsafe.palletsprojects.com/en/2.1.x/escaping/#markupsafe.Markup)). That could be considered lame though :)

Yes, please add the documentation, referencing these Jinja pages. I think a good place could be the Design document, not the User Guide, since this is mostly internal to StrictDoc. Add a section right after this one: https://strictdoc.readthedocs.io/en/stable/strictdoc_25_design.html#parsing-sdoc-files?

Reply:
> What is the immediate member?

```
from jinja2 import Template

# Define a template string
template_string = """
{%- set field_content_ = foo2|safe + foo|safe -%}
{{ field_content_ }}
"""

# Create a template object
template = Template(template_string, autoescape=False)

# Render the template with the context
rendered_html = template.render(foo="FOO", foo2="<div>TEST</div>")

# Print the rendered HTML
print(rendered_html)
# <div>TEST</div>FOO
```

If you remove the first `safe`, then the escaped string is printed:

```
&lt;div&gt;TEST&lt;/div&gt;FOO
```

In your line, it is only the 
".&nbsp;" that is made to Markup, and that makes the whole expression to go escaped. This is exactly the behavior that is quite confusing when you have to read it right-to-left and do the mental magic. 

```
      {# WIP: DO NOT UNDERSTAND WHY adding |safe here makes the section title to be escaped #}
      {%- set field_content_ = field_content_ + sdoc_entity.context.title_number_string + ".&nbsp;"|safe -%}
```

Reply:
>> Could you explain what you mean with "both directions"?
>
> The order of Markup and str doesn't matter.

IIUC it works by defining the [`__radd__`](https://docs.python.org/3/reference/datamodel.html#object.__radd__) method.

> These methods are called to implement the binary arithmetic operations (+, -, *, @, /, //, %, [divmod()](https://docs.python.org/3/library/functions.html#divmod), [pow()](https://docs.python.org/3/library/functions.html#pow), **, <<, >>, &, ^, |) with reflected (swapped) operands.

See [Markup.__radd__](https://github.com/pallets/markupsafe/blob/main/src/markupsafe/__init__.py#L139).

Reply:
Yes, this is how it works! Super-clear now.